### DISCUSSION TITLE
Time limit for a series of survey text trials

### DISCUSSION THREAD

**Discussion by michelleramey** [jspsych/jsPsych#2310] [Q&A] [answered]
Hello!
I am trying to make an experiment that asks participants to think of as many answers to a prompt as they can within 1 minute (using survey-text). I am looking to record the response as well as the rt (or clock time) of each response.
I imagine this would involve allowing a dynamic number (or perhaps an excessively large number) of survey-text trials with a single text entry field for each one. Each time a participant submits a response, a new survey-text field would appear, until 1 minute has elapsed, at which point the timeline would end.

I've pored over past discussions and I haven't been able to find something like this. Is it possible?

Thanks in advance!

-----

Here's what I've got so far (with 5s limit for testing), but I've unsuccessfully tried a variety of ways to get it to record the responses on time out. I also don't think this strategy would allow me to record the rt of each response. (Insert code function is not working properly for this so apologies for bad formatting)
```js
var time_limit = 5000;
var start_time2;
var end_test_timer;
var trial_count = 0;
var n_trials = 20;
var counter2 = 1;

var trial = {
    timeline: [ {
        timeline: [
            {type: jsPsychSurveyText,
                questions: [
        {prompt: 'Probe', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40},
        {prompt: ' ', placeholder: 'Type here', columns: 40}

      ],
      randomize_question_order: false,
            on_load: function() {
                trial_count++;
                if (trial_count > 0) {
                    start_time2 = performance.now();
                    var end_test_timer = setTimeout(function() {
                        var end_time = performance.now();
                        var elapsed_time = end_time - start_time2;
                        console.log("elapsed time: ", elapsed_time);
     
                        jsPsych.endCurrentTimeline();
                        jsPsych.finishTrial({status: "ended early"});
                    }, time_limit);
                }
            }, 
            on_finish: function() {
                counter2++;
                if (trial_count == n_trials) {
                    clearTimeout(end_test_timer);
                }
                    }
            }
        ], 
    }]
};

    jsPsych.run([trial]);
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @michelleramey, great question. Do you need all of the text boxes to appear on the same page? Or one textbox per page, and the participant submits their response (key press or button click), which triggers the next textbox to appear? If the latter method is ok, my approach to this would be to have one survey-text trial that is inside of a loop node. Then in the loop function, you can check whether the time limit is up - if so then stop the loop, otherwise repeat the timeline (i.e. run another survey-text trial). The advantage of this showing one textbox per page is that it'll give you RTs for each response submission.

Let us know if that approach sounds like it would work, and if you want any further guidance with the code.

**Reply by michelleramey** [reply]
Hi Becky, thank you so much for your help! I don't need the textboxes to appear on one page, so your suggested method with one textbox per page sounds perfect. I've never used loop nodes before but I will take a crack at it. 
Thank you again!

**Reply by becky-gilbert** [reply]
Great! Here's the [documentation on looping timelines](https://www.jspsych.org/7.0/overview/timeline/#looping-timelines). 

One thing I realized is that the loop node will stop the trials from continuing after the time limit has been exceeded, but you probably also want to immediately end the current trial when the time limit is reached. To do that, I would set up a timer that starts right before the first survey-text trial, and when the time limit is reached, it ends the current trial (while also saving the response for that trial, if there is one). I would do it like this:
```js
// initialize global variable - change this to true when the time is up, and check its value in the loop function
var timed_out = false; 

// right before your first survey-text trial, set up a timer
var trial_before_survey_text = {
  // ... other parameters ...
  on_finish: function() {
    // this timer will run the function below when the 5s time limit is reached
    jsPsych.pluginAPI.setTimeout(function() {
      // get the current textbox response, if there is one
      var response = null;
      var textbox = document.getElementsByTagName("input")[0];
      if (textbox) {
        response = textbox.value;
      }
      timed_out = true;
      // finish the current survey-text trial and save the data
      jsPsych.finishTrial({response: response, status: "ended_early"});
    }, 5000);
  }
};

// set up the survey-text trial ...
var trial = {
  type: jsPsychSurveyText
  // ...
};

// set up the loop node to keep repeating the trial until timed_out is true
var survey_text_loop = {
  timeline: [trial],
  loop_function: function() {
    if (timed_out) {
      return false;
    } else {
      return true;
    }
  }
};
```
Does that work for you?
